Feature/Ja-38-Login

GrzegorzKmiecinski opened the pull request:
Created Backend for login feature: new method in DataAccess that checks if input data for user login is correct and few new methods in DataAccess that return information about currently logged in user.

GrzegorzKmiecinski committed:
Create backend for Login feature

Created method that check if input data for user login is correct. Created few methods that return information about currently logged in user.

Zjyslav commented on TutorLizard.BusinessLogic/Data/DataAccess.cs:
```diff
@@ -16,6 +16,8 @@ public class DataAccess : IUserIdentityDataAccess, IStudentDataAccess, ITutorDat
     private string _adRequestFilePath = @"Data/ad_requests.json";
     private string _scheduleItemRequestFilePath = @"Data/schedule_requests.json";
 
+    private User _currentUser = new();
```
I think that this is more of `UserIdentityService`'s responsibility to keep track of current user. I'd move this functionality there, because it's not strictly connected with reading and writing data.

Zjyslav commented on TutorLizard.BusinessLogic/Data/DataAccess.cs:
```diff
@@ -196,4 +198,55 @@ private int GetNewScheduleItemRequestID()
     }
     #endregion
+
+    private bool DoesTheUserIdExist(int id)
```
This relies too heavily on assumption that we'll never have any gaps in ids. Any manual deletion could break it. Why not simple `Any()`?

GrzegorzKmiecinski committed:
Moved some functionality from DataAccess to UserIdentityService

Also created filled some methods with code in UserIdentityService.  Changed how we get id for Get[]ID methods in DataAccess to return 1 instead of 0 if the file is empty. Still missing Log out method.

GrzegorzKmiecinski committed:
fixed minor issue

Zjyslav commented on TutorLizard.BusinessLogic/Data/DataAccess.cs:
```diff
@@ -159,41 +164,82 @@ public ScheduleItemRequest CreateScheduleItemRequest (int scheduleItemId, int us
+
+        return false;
+    }
+    private User ReturnActiveUser(int id)
```
There are other cases for looking up users by id except active user. It will be called each time we want to show a user's name (displaying ad, reviewing requests).
I think calling it something like `GetUserById` would be better.

Zjyslav commented on TutorLizard.BusinessLogic/Data/DataAccess.cs:
```diff
@@ -159,41 +164,82 @@ public ScheduleItemRequest CreateScheduleItemRequest (int scheduleItemId, int us
+    }
+    private User ReturnActiveUser(int id)
+    {
+        User tempUser = _userList.Single(x => x.Id == id);
```
Single will throw exception if not found. If you return SingleOrDefault, it will return null when not found. You'd have to change return type of the method to nullable (User?).

Zjyslav commented on TutorLizard.BusinessLogic/Services/UserIdentityService.cs:
```diff
@@ -5,47 +5,76 @@ namespace TutorLizard.BusinessLogic.Services;
+
+    public User? GetActiveUser() 
+    {
+        if (_activeUser is null)
```
If you don't do this check and `_activeUser' is null, it will return null either way.

Zjyslav commented on TutorLizard.BusinessLogic/Services/UserIdentityService.cs:
```diff
@@ -5,47 +5,76 @@ namespace TutorLizard.BusinessLogic.Services;
         _dataAccess = dataAccess;
     }
+
+    public User? GetActiveUser() 
```
I didn't put it in the interface on purpose, because I wanted to limit exposure of user data in UI (UI doesn't have access to DataAccess). If we expand User type by some sensitive data, e.g. password, putting the whole User object could potentially allow access it.
It doesn't really hurt us, so it can stay, but it's something to think about

Zjyslav commented on TutorLizard.BusinessLogic/Services/UserIdentityService.cs:
```diff
@@ -5,47 +5,76 @@ namespace TutorLizard.BusinessLogic.Services;
 public class UserIdentityService : IUserIdentityService
 {
     private readonly IUserIdentityDataAccess _dataAccess;
+    private readonly DataAccess? _dataAccessObject = new();
```
You already have a `IUserIdentityDataAccess` object in this class. You don't need this.
We create (actually receive through constructor) DataAccess as an interface, so that this class is not dependent on any given implementation. If you need access to any of `DataAccess`'s methods that are not in the interface, simply expand the interface.

Zjyslav commented on TutorLizard.BusinessLogic/Services/UserIdentityService.cs:
```diff
@@ -5,47 +5,76 @@ namespace TutorLizard.BusinessLogic.Services;
-        // return id of newly created user
-        // return 0 if unsuccessful
-        throw new NotImplementedException();
+        if (!IsUserNameTaken(userName))
```
If it's not taken, then it's available and should be successful.

Zjyslav commented on TutorLizard.BusinessLogic/Services/UserIdentityService.cs:
```diff
@@ -5,47 +5,76 @@ namespace TutorLizard.BusinessLogic.Services;
+        return newUser.Id;
+    }
+
+    public void ResetCurrentUserToNull()
```
This is exactly what `LogOut()` method should be doing.

Zjyslav commented on TutorLizard.BusinessLogic/Data/DataAccess.cs:
```diff
@@ -159,41 +164,82 @@ public ScheduleItemRequest CreateScheduleItemRequest (int scheduleItemId, int us
+        return (first: true, last: ReturnActiveUser(id));
+    }
+
+    public bool LookForUserName(string username)
```
Name is a little confusing. Look for suggests to me that you return what you find, not info that it exists.
However, it works fine, so it could stay that way.

Zjyslav commented on TutorLizard.BusinessLogic/Data/DataAccess.cs:
```diff
@@ -159,41 +164,82 @@ public ScheduleItemRequest CreateScheduleItemRequest (int scheduleItemId, int us
     private int GetNewUserID()
     {
-        if (_userList.Count() > 0)
+        if (_userList.Count() +1 > 1)
```
It does the same thing as before the change, but with one more calculation to me made.
Using `Any()` would be more efficient.

Zjyslav reviewed (changes requested):
All previously requested changes have been made.
Some of my new comments is just nitpicking and can be ignored, but please use `_dataAccess` object that's already in UserIdentityService instead of a new `DataAccess`.

GrzegorzKmiecinski committed:
Requested changes implemented.

Merged with develop. Removed DataAccess Obj from UserIdentityService. Expanded the UserIdentityDataAccess Interface.

GrzegorzKmiecinski committed:
Change RegisterUser method

GrzegorzKmiecinski committed:
fix issue with interface

GrzegorzKmiecinski committed:
Change Get ID methods

GrzegorzKmiecinski committed:
Fix an issue with DoesTheUsernameMatchUserId method

GrzegorzKmiecinski committed:
fix minor issue

GrzegorzKmiecinski committed:
Implement requested changes
